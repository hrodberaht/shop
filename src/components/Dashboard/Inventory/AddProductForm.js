import React, { Component } from 'react';
import { Field, reduxForm } from 'redux-form';
import PropTypes from 'prop-types';
import * as yup from 'yup';
import { connect } from 'react-redux';

const schema = yup.object().shape({
  name: yup.string().required(),
  type: yup.string().required(),
  price: yup.number().required(),
  inStock: yup.number().required(),
});

const asyncValidate = values => schema.validate(values, { abortEarly: false }).catch((err) => {
  const errorsForm = {};
  err.inner.forEach((error) => {
    errorsForm[error.path] = error.type;
  });
  throw errorsForm;
});

export class AddProductForm extends Component {
  renderField = ({ input, type, meta: { touched, error } }) => (
    <div>
      <div>
        <input className="form-input" {...input} type={type} />
        {touched && error && <p className="error-text">{error}</p>}
      </div>
    </div>
  );

  render() {
    const { handleSubmit, pristine } = this.props;

    return (
      <div className="product-form">
        <form onSubmit={handleSubmit}>
          <label htmlFor="name">
            <span>Name:</span>
            <Field id="name" name="name" component={this.renderField} type="text" />
          </label>
          <label htmlFor="type">
            <span>Type:</span>
            <Field id="type" name="type" component={this.renderField} type="text" />
          </label>
          <label htmlFor="price">
            <span>Price:</span>
            <Field id="price" name="price" component={this.renderField} type="number" />
          </label>
          <label htmlFor="inStock">
            <span>Stock:</span>
            <Field id="inStock" name="inStock" component={this.renderField} type="number" />
          </label>
          <button className="btn btn-submit" type="submit" disabled={pristine}>
            Add
          </button>
        </form>
      </div>
    );
  }
}

const mapStateToProps = (state, ownProps) => ({
  initialValues: ownProps.product,
  form: ownProps.product.id,
});

export default connect(mapStateToProps)(
  reduxForm({
    asyncValidate,
    enableReinitialize: true,
    form: 'addProduct',
  })(AddProductForm),
);

AddProductForm.propTypes = {
  handleSubmit: PropTypes.func.isRequired,
};
