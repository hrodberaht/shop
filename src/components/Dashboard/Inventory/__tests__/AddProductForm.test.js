import { AddProductForm } from '../AddProductForm';

const defaultProps = {
  handleSubmit: jest.fn(),
};

const setup = buildSetup(AddProductForm, defaultProps);

describe('<AddProductForm />', () => {
  it('should render without crash', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });
});
