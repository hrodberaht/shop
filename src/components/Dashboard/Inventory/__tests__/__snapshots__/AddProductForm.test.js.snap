// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddProductForm /> should render without crash 1`] = `
<div
  className="product-form"
>
  <form
    onSubmit={[MockFunction]}
  >
    <label
      htmlFor="name"
    >
      <p>
        Name:
      </p>
      <Field
        component={[Function]}
        id="name"
        name="name"
        type="text"
      />
    </label>
    <label
      htmlFor="type"
    >
      <p>
        Type:
      </p>
      <Field
        component={[Function]}
        id="type"
        name="type"
        type="text"
      />
    </label>
    <label
      htmlFor="price"
    >
      <p>
        Price:
      </p>
      <Field
        component={[Function]}
        id="price"
        name="price"
        type="number"
      />
    </label>
    <label
      htmlFor="inStock"
    >
      <p>
        Stock:
      </p>
      <Field
        component={[Function]}
        id="inStock"
        name="inStock"
        type="number"
      />
    </label>
    <button
      className="btn btn-submit"
      type="submit"
    >
      Add
    </button>
  </form>
</div>
`;
